src/main/java/com/vmware/vipclient/i18n/messages/service/ProductService.java

```diff
@@ -72,44 +73,75 @@ public ProductService(BaseDTO dto) {
         return components;
     }
 
+    private Set<String> getSupportedLocales() {
+        return getSupportedLocales(true);
+    }
+
+    public Set<String> getCachedSupportedLocales() {
```

Comment:
Please review the other issue linked to this: https://github.com/vmware/singleton/issues/877: 

We do not need getCachedXXX for other data.

getCachedSupportedLcoales is an exception because we use it to pre-determine if getComponentMessages must even proceed or not. This was done only for performance purposes, so that if locale is not supported, it won't even try to fetch the bundle and directly serve fallback locale bundle. However, if getSupportedLocales was not previously cached or if it previously failed, it should not block getComponentMessages. The client should just directly send  the call to getComponentMessages.

Reply:
It seems you created so much code just to work around an extreme situation, getSupportedLocales keeps fail. In this case,  we should remove test supported locales because stability takes precendence over performance. Let users take charge of it. Get the supported locale before getting translation or bear the delay of unsupported locale. In fact this logic has been there before for a long time.